include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -825,6 +825,70 @@ struct __gen_unique_mask
     _BinaryPredicate __pred;
 };
 
+template <typename _IsOpDifference, typename _Compare>
+struct __gen_set_mask
+{
+    template <typename _InRng>
+    bool
+    operator()(const _InRng& __in_rng, std::size_t __id) const
+    {
+        // First we must extract individual sequences from zip iterator because they may not have the same length,
+        // dereferencing is dangerous
+        auto __set_a = std::get<0>(__in_rng.tuple());    // first sequence
+        auto __set_b = std::get<1>(__in_rng.tuple());    // second sequence
+        auto __set_mask = std::get<2>(__in_rng.tuple()); // mask sequence
+
+        std::size_t __nb = __set_b.size();
+
+        auto __val_a = __set_a[__id];
+
+        auto __res = oneapi::dpl::__internal::__pstl_lower_bound(__set_b, std::size_t{0}, __nb, __val_a, __comp);
+
+        bool bres =
+            _IsOpDifference::value; //initialization in true in case of difference operation; false - intersection.
+        if (__res == __nb || __comp(__val_a, __set_b[__res]))
+        {
+            // there is no __val_a in __set_b, so __set_b in the difference {__set_a}/{__set_b};
+        }
+        else
+        {
+            auto __val_b = __set_b[__res];
+
+            //Difference operation logic: if number of duplication in __set_a on left side from __id > total number of
+            //duplication in __set_b than a mask is 1
+
+            //Intersection operation logic: if number of duplication in __set_a on left side from __id <= total number of
+            //duplication in __set_b than a mask is 1
+
+            const std::size_t __count_a_left =
+                __id - oneapi::dpl::__internal::__pstl_left_bound(__set_a, std::size_t{0}, __id, __val_a, __comp) + 1;
+
+            const std::size_t __count_b =
+                oneapi::dpl::__internal::__pstl_right_bound(__set_b, __res, __nb, __val_b, __comp) - __res + __res -
```

Comment:
I don't understand the ` - __res + __res` portion of this calculation. Is it necessary? If so, can we add a comment explaining it?

Reply:
This is me propagating an benign error from the multi-pass scan implementation which inspired this one.  I think they can just be removed in both places, good catch.